src/lib.rs

```diff
@@ -935,22 +896,19 @@ impl Drop for Rav1dContext {
         // remove all pointers from the structure. We can't make the drop
         // function unsafe because the Drop trait requires a safe function.
         unsafe {
-            if !(self.tc).is_null() {
+            if self.n_tc > 1 {
                 let ttd: &TaskThreadData = &*self.task_thread;
-                if self.n_tc > 1 {
-                    let task_thread_lock = ttd.delayed_fg.lock().unwrap();
-                    for tc in &self.tc_shared {
-                        tc.thread_data.die.store(true, Ordering::Relaxed);
-                    }
-                    ttd.cond.notify_all();
-                    drop(task_thread_lock);
-                    for tc in &self.tc_shared {
-                        if let Some(handle) = tc.handle {
-                            pthread_join(handle, 0 as *mut *mut c_void);
-                        }
+                let task_thread_lock = ttd.delayed_fg.lock().unwrap();
+                for tc in self.tc.iter() {
+                    tc.thread_data.die.store(true, Ordering::Relaxed);
+                }
+                ttd.cond.notify_all();
+                drop(task_thread_lock);
+                for task_thread in self.tc.iter_mut() {
+                    if let Some(handle) = task_thread.handle.take() {
+                        handle.join().expect("Could not join task thread");
                     }
                 }
-                rav1d_free_aligned(self.tc as *mut c_void);
             }
             let mut n_1: c_uint = 0 as c_int as c_uint;
             while !(self.fc).is_null() && n_1 < self.n_fc {
```

Comment:
I realized this too late (not on #717), and can't comment on what's below since it's too far from the diff, but all of the `mem::take`s below can be removed now that this is in an `impl Drop` since they'll be done automatically.

Reply:
We can rewrite Drop in a separate PR, I figured we'd do that when it's completely safe, as those `take`s aren't doing anything bad and just cost a tiny amount.